google-cloud-spanner/src/test/java/com/google/cloud/spanner/TransactionChannelHintTest.java

```diff
@@ -0,0 +1,335 @@
+
+  @AfterClass
+  public static void resetLogging() {
+    Logger logger = Logger.getLogger("");
+    logger.setLevel(originalLogLevel);
+  }
+
+  @After
+  public void reset() {
+    mockSpanner.reset();
+    executeSqlLocalIps.clear();
+    streamingReadLocalIps.clear();
+  }
+
+  private SpannerOptions createSpannerOptions() {
+    String endpoint = address.getHostString() + ":" + server.getPort();
+    return SpannerOptions.newBuilder()
+        .setProjectId("[PROJECT]")
+        .setChannelConfigurator(
+            input -> {
+              input.usePlaintext();
+              return input;
+            })
+        .setCompressorName("gzip")
+        .setHost("http://" + endpoint)
+        .setCredentials(NoCredentials.getInstance())
+        .build();
+  }
+
+  @Test
+  public void testSingleUse_usesSingleChannel() {
+    try (Spanner spanner = createSpannerOptions().getService()) {
+      DatabaseClient client = spanner.getDatabaseClient(DatabaseId.of("p", "i", "d"));
+      try (ResultSet resultSet = client.singleUse().executeQuery(SELECT1)) {
+        while (resultSet.next()) {}
+      }
+    }
+    assertEquals(1, executeSqlLocalIps.size());
+  }
+
+  @Test
+  public void testSingleUse_withTimestampBound_usesSingleChannel() {
+    try (Spanner spanner = createSpannerOptions().getService()) {
+      DatabaseClient client = spanner.getDatabaseClient(DatabaseId.of("p", "i", "d"));
+      try (ResultSet resultSet =
+          client
+              .singleUse(TimestampBound.ofExactStaleness(15L, TimeUnit.SECONDS))
+              .executeQuery(SELECT1)) {
+        while (resultSet.next()) {}
+      }
+    }
+    assertEquals(1, executeSqlLocalIps.size());
+  }
+
+  @Test
+  public void testSingleUseReadOnlyTransaction_usesSingleChannel() {
+    try (Spanner spanner = createSpannerOptions().getService()) {
+      DatabaseClient client = spanner.getDatabaseClient(DatabaseId.of("p", "i", "d"));
+      try (ResultSet resultSet = client.singleUseReadOnlyTransaction().executeQuery(SELECT1)) {
+        while (resultSet.next()) {}
+      }
+    }
+    assertEquals(1, executeSqlLocalIps.size());
+  }
+
+  @Test
+  public void testSingleUseReadOnlyTransaction_withTimestampBound_usesSingleChannel() {
+    try (Spanner spanner = createSpannerOptions().getService()) {
+      DatabaseClient client = spanner.getDatabaseClient(DatabaseId.of("p", "i", "d"));
+      try (ResultSet resultSet =
+          client
+              .singleUseReadOnlyTransaction(TimestampBound.ofExactStaleness(15L, TimeUnit.SECONDS))
+              .executeQuery(SELECT1)) {
+        while (resultSet.next()) {}
+      }
+    }
+    assertEquals(1, executeSqlLocalIps.size());
+  }
+
+  @Test
+  public void testReadOnlyTransaction_usesSingleChannel() {
+    try (Spanner spanner = createSpannerOptions().getService()) {
+      DatabaseClient client = spanner.getDatabaseClient(DatabaseId.of("p", "i", "d"));
+      try (ReadOnlyTransaction transaction = client.readOnlyTransaction()) {
```

Comment:
Read-only transactions (currently) use an explicit `BeginTransaction` RPC. Can we verify that not only the `ExecuteStreamingSql` calls all use the same channel, but that also the `BeginTransaction` RPC uses the same channel?

Reply:
Done